testunit.go

```diff
@@ -10,102 +9,88 @@ import (
+	//Case name
+	Name string
+	//Args is used to generate bundle
+	Args string
+	//Describle what does this unit test for. It is optional.
+	Description string
 
-	myruntime, err := factory.CreateRuntime(Runtime)
-	if err != nil {
-		logrus.Printf("Create runtime %v err: %v\n", Runtime, err)
-	}
-	if myruntime.NeedConvert() {
-		if _, err := myruntime.Convert(validateObj, TestCacheDir); err != nil {
-			logrus.Fatalln(err)
-		}
-	}
+	BundleDir string
+	Runtime   factory.Factory
+}
 
-	testopt := utils.GetAfterNStr(configArgs, "--args=./runtimetest --args=", 3)
-	testRoot := path.Join(TestCacheDir, validateObj)
-	out, err := myruntime.StartRT(testRoot)
-	if err != nil {
-		//logrus.Printf("Run runtime err: %v\n", err)
-		return err
+func LoadTestUnits(filename string) (units []TestUnit) {
+	for key, value := range config.BundleMap {
+		//TODO: config.BundleMap should support 'Description'
+		unit := TestUnitNew(key, value, "")
+		units = append(units, unit)
 	}
-	switch testopt {
-	case "vna":
-		err = hooks.SetPostStartHooks(out, hooks.NamespacePostStart)
-	default:
-	}
-
-	return nil
+	return units
 }
 
-func generateConfigs(validateObj string, configArgs string) {
-	args := splitArgs(configArgs)
+func TestUnitNew(name string, args string, desc string) (unit TestUnit) {
+	unit.Name = name
+	unit.Args = args
+	unit.Description = desc
 
-	logrus.Debugf("Args to the ocitools generate: ")
-	for _, a := range args {
-		logrus.Debugln(a)
-	}
-	Mutex.Lock()
-	_, err := utils.ExecGenCmd(args)
-	if err != nil {
-		logrus.Fatalf("Generate *.json err: %v\n", err)
-	}
-
-	copy("./plugins/runtime.json-"+validateObj, "./plugins/runtime.json")
-	if err != nil {
-		logrus.Fatalf("copy to runtime.json-%v, %v", validateObj, err)
-	}
+	return unit
+}
 
-	copy("./plugins/config.json-"+validateObj, "./plugins/config.json")
-	if err != nil {
-		logrus.Fatalf("copy to config.json-%v, %v", validateObj, err)
+//Set runtime
+func (unit *TestUnit) SetRuntime(runtime string) error {
+	if r, err := factory.CreateRuntime(runtime); err != nil {
+		logrus.Printf("Create runtime %v err: %v\n", runtime, err)
+		return err
+	} else {
+		unit.Runtime = r
 	}
-	Mutex.Unlock()
+	return nil
 }
 
-func splitArgs(args string) []string {
-
-	argsnew := strings.TrimSpace(args)
-
-	argArray := strings.Split(argsnew, "--")
+func (unit *TestUnit) Run() error {
+	if unit.Runtime == nil {
+		logrus.Fatalf("Set the runtime before run the test")
+	}
 
-	lenth := len(argArray)
-	resArray := make([]string, lenth-1)
-	for i, arg := range argArray {
-		if i == 0 || i == lenth {
-			continue
-		} else {
-			resArray[i-1] = "--" + strings.TrimSpace(arg)
-		}
+	unit.GenerateConfigs()
+	unit.PrepareBundle()
+	_, err := unit.Runtime.StartRT(unit.BundleDir)
+	return err
+	/* TODO: what is testopt used for ?
+	testopt := utils.GetAfterNStr(configArgs, "--args=./runtimetest --args=", 3)
+	switch testopt {
+	case "vna":
+		err = hooks.SetPostStartHooks(out, hooks.NamespacePostStart)
+	default:
 	}
```

Comment:
@liangchenye `testopt` is the `specs` item to be test , which is the argument after `./runtimetest`, for some item we should add hooks before or after  we start container runtime